From 65985740cc15824fb401d1e640280d00c2faf384 Mon Sep 17 00:00:00 2001
From: Johnny Mnemonic <jm@machine-hall.org>
Date: Sun, 12 Jan 2025 21:30:13 +0100
Subject: [PATCH 24/25] Introduce __thread_pointer() also for ia64

Related to "Move <thread_pointer.h> to kernel-independent sysdeps
directories"

See 7a3e2e877a70153a6d1b786925b34f3b396e20f1 for reference.
---
 sysdeps/ia64/nptl/tls.h       |  6 +++---
 sysdeps/ia64/thread_pointer.h | 31 +++++++++++++++++++++++++++++++
 2 files changed, 34 insertions(+), 3 deletions(-)
 create mode 100644 sysdeps/ia64/thread_pointer.h

diff --git a/sysdeps/ia64/nptl/tls.h b/sysdeps/ia64/nptl/tls.h
index dbfb85a7f8..f70aabc699 100644
--- a/sysdeps/ia64/nptl/tls.h
+++ b/sysdeps/ia64/nptl/tls.h
@@ -1,5 +1,5 @@
 /* Definition for thread-local data handling.  nptl/IA-64 version.
-   Copyright (C) 2003-2024 Free Software Foundation, Inc.
+   Copyright (C) 2003-2025 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
 
    The GNU C Library is free software; you can redistribute it and/or
@@ -27,6 +27,8 @@
 # include <stdlib.h>
 # include <list.h>
 # include <dl-dtv.h>
+// for __thread_self
+# include <thread_pointer.h>
 
 typedef struct
 {
@@ -34,8 +36,6 @@ typedef struct
   void *__private;
 } tcbhead_t;
 
-register struct pthread *__thread_self __asm__("r13");
-
 # define TLS_MULTIPLE_THREADS_IN_TCB 1
 
 #else /* __ASSEMBLER__ */
diff --git a/sysdeps/ia64/thread_pointer.h b/sysdeps/ia64/thread_pointer.h
new file mode 100644
index 0000000000..a7e15734cd
--- /dev/null
+++ b/sysdeps/ia64/thread_pointer.h
@@ -0,0 +1,31 @@
+/* __thread_pointer definition.  ia64 version.
+   Based on powerpc version.
+   Copyright (C) 2025 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library.  If not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#ifndef _SYS_THREAD_POINTER_H
+#define _SYS_THREAD_POINTER_H
+
+register struct pthread *__thread_self __asm__("r13");
+
+static inline void *
+__thread_pointer (void)
+{
+  return __thread_self;
+}
+
+#endif /* _SYS_THREAD_POINTER_H */
-- 
2.25.1

